### ISSUE TITLE
C_OpenSession deadlocks if C_INITIALIZE(.flags=CKF_OS_LOCKING_OK) is called

### ISSUE THREAD

**Issue by PeterHuewe** [tpm2-software/tpm2-pkcs11#38]
C_OpenSession ends with a deadlock if C_INITIALIZE is called with .flags=CKF_OS_LOCKING_OK

Steps to reproduce:
1) Pass CKF_OS_LOCKING_OK flag to C_INITIALIZE
2) Call OpenSession
If compiled with --debug:
 _pkcs-open-close-session.int: src/lib/mutex.h:70: mutex_lock_fatal: Assertion `rv == CKR_OK' failed._

See example  below (modified from test/integration/pkcs-open-close-session.int.c)

```

 CK_C_INITIALIZE_ARGS init_args ={0};
 init_args.flags=2;

 CK_RV rv = C_Initialize(&init_args);
    if(rv != CKR_OK){
        LOGE("C_Initialize failed! Response Code %x", rv);
        exit(1);
    }


    CK_SESSION_HANDLE handle;

    rv = C_OpenSession(0x01, CKF_SERIAL_SESSION | CKF_RW_SESSION, NULL , NULL, &handle);
// We never get here!
// pkcs-open-close-session.int: src/lib/mutex.h:70: mutex_lock_fatal: Assertion `rv == CKR_OK' failed.
// Abgebrochen

    if(rv != CKR_OK){
        LOGE("C_OpenSession failed! Response Code %x", rv);
        exit(1);
    }

```

**Comment by PeterHuewe** [comment]
Okay the problem lies in the following order:
src/lib/session.c , session_open line 70: 
        `session_table_lock(global.s_table);`
session_table_lock now holds global.s_table mutex
line 71:
        `rv = check_max_sessions(is_rw);`
which calls 
```
static CK_RV check_max_sessions(bool is_rw) {

    unsigned long cnt = session_table_get_cnt(global.s_table, is_rw);
    return (cnt > MAX_NUM_OF_SESSIONS) ?
        CKR_SESSION_COUNT : CKR_OK;
}
```

which calls 
```
unsigned long session_table_get_cnt(session_table *t, bool is_rw) {
    session_table_lock(t);
    unsigned long tmp = is_rw ? t->rw_cnt : t->cnt;
    session_table_unlock(t);
    return tmp;
}
```

--> session_table_get_cnt wants to take the mutex again on global.s_table (which is already held)
-> deadlock.


Question is whether the locking in session_open can simply be removed?

**Comment by williamcroberts** [comment]
This seems to be fixed in #47 and #48
